### PR TITLE
musesampler_overlapping_notes

### PR DESCRIPTION
**Pull request by RomanPudashkin**
Resolves: https://github.com/musescore/MuseScore/issues/15306

### CODE REVIEW

**Comment by Jojo-Schmitz** [comment]
Hmm, shouldn't that rather be a track for each **voice** (incidently called track internally)?

**Comment by cbjeukendrup** [comment]
@Jojo-Schmitz It looks like it is more complicated... This is what I understand of it by reading the code of this PR: (@RomanPudashkin please correct me if I'm wrong)

`engraving`'s model: just a list of "tracks", where "staff index" and "voice" are completely determined by these tracks with the formula `track = staffIdx * VOICES + voiceIdx`

MuseSampler's model: apparently within one MuseSampler instance there is a concept of "tracks", and for events on these "tracks" you can additionally specify a "voice index". So, a MuseSampler instance has multiple tracks, and each track can contain multiple voices.

This results in the following mapping:
- `engraving` staff index -> MuseSampler track
- `engraving` voice -> MuseSampler voice

It's a bit overwhelming that the word "track" seems to have at least three different meanings now (one in `engraving`, one in the Mixer, and one in MuseSampler), but that seems to be just the way it is.

(Personally I had also expected that we would just abuse MuseSampler's voice API by specifying `staffIndex * VOICES + voiceIndex`, but apparently MuseSampler has an extra concept for different staves.)

**Comment by cbjeukendrup** [comment]
@RomanPudashkin I still have some doubts about that optimisation of doing this for keyboards only:
- Having multiple staves for non-keyboard instruments is very uncommon, so this optimisation won't have any effect for 99,95% of scores (uneducated estimation)
- In those cases where someone _still_ uses multiple staves in a non-keyboard instrument, overlapping notes are actually even _more_ probable than in keyboard music: in keyboard music, generally one staff is for the high range and the other for the low range, but in non-keyboard music both staves are likely to cover the same range. Consider for example the case where a violin section has to play _divisi_, divided into two subsections; naturally, both staves will play in a similar range (since for violin you wouldn't really use any other clef than G and the common range is not huge) so overlapping notes between these staves are actually more common and thus applying the fix to non-keyboard staves is actually even _more_ worthwhile than for keyboard staves.

**Comment by basaigh** [comment]
As a harpist, the two staves are quite often used to denote what each hand plays, not always having anything to do with pitch ranges. The issue of the same note being played twice is also particularly prevalent on harp as once you pluck/strike the string the note will hold without interaction, leaving the string free for either hand to use it immediately.
I've been watching the attached issue for quite a while now, and would really appreciate if the fix was not limited to keyboard instruments.

**Comment by MarcSabatella** [comment]
I agree - while multiple staves not *common* for non-keyboard instruments (well, besides harp, and marimba), when it does occur, it’s that much more likely to contain overlaps.  Divisi parts are an obvious use case.  Another might be SATB closed-score choral music if there is ever one generic “voice” sound covering the full range, or if the user substitutes something else.  If an instrument has multiple staves for any reason, this is still a potential issue.

But - thank you *so* much for tackling this!

**Comment by dcorson-ticino-com** [comment]
I would say that multiple staves ARE common for divided strings.
I play viola in the local orchestra and per happenstance there are divided strings with multiple staves in all the 3 pieces we are playing in our next concert: Mendelssohn, Boccherini and Rossini.

File: src/framework/musesampler/internal/musesamplersequencer.cpp [hunk 1]
```diff
@@ -79,13 +79,16 @@ static const std::unordered_map<mpe::ArticulationType, ms_NoteArticulation> ARTI
...
-void MuseSamplerSequencer::init(MuseSamplerLibHandlerPtr samplerLib, ms_MuseSampler sampler, IMuseSamplerTracksPtr tracks,
-                                std::string&& defaultPresetCode)
+void MuseSamplerSequencer::init(const mpe::PlaybackSetupData& setupData, const std::string& defaultPresetCode,
+                                MuseSamplerLibHandlerPtr samplerLib, ms_MuseSampler sampler, IMuseSamplerTracksPtr tracks)
 {
-    m_samplerLib = std::move(samplerLib);
-    m_sampler = std::move(sampler);
-    m_track = std::move(track);
-    m_defaultPresetCode = std::move(defaultPresetCode);
+    // Optimization: don't add new tracks for non-keaboards
```
Comment (Jojo-Schmitz) [inline src/framework/musesampler/internal/musesamplersequencer.cpp:85]: typo: non-ke__y__boards

**Comment by DmitryArefiev** [comment]
Retested on Win10, Mac13.6 with **musesampler 0.6.0.104** (and MuseSounds). 
#15306 looks good on keys, Harp and Strings with divisi parts. 
Scores with a lot of additional staves now took more RAM and CPU (when using musesampler v6) but it's expected and not critical

**Review by cbjeukendrup — APPROVED** [review]

**Pull request merged**